jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/EcrUtils.kt

```diff
@@ -0,0 +1,100 @@
+import com.intellij.openapi.project.Project
+import com.intellij.remoteServer.impl.configuration.RemoteServerImpl
+import com.intellij.remoteServer.runtime.ServerConnectionManager
+import com.intellij.remoteServer.runtime.ServerConnector
+import com.intellij.remoteServer.runtime.deployment.ServerRuntimeInstance
+import com.intellij.util.Base64
+import kotlinx.coroutines.future.await
+import org.jetbrains.concurrency.AsyncPromise
+import org.jetbrains.concurrency.await
+import software.amazon.awssdk.services.ecr.model.AuthorizationData
+import software.aws.toolkits.jetbrains.services.ecr.actions.LocalImage
+import software.aws.toolkits.jetbrains.services.ecr.resources.Repository
+
+data class EcrLogin(
+    val username: String,
+    val password: String
+)
+
+data class EcrPushRequest(
+    val localImageId: String,
+    val remoteRepo: Repository,
+    val remoteTag: String
+)
+
+suspend fun getDockerServerRuntimeInstance(): DockerServerRuntimeInstance {
+    val instancePromise = AsyncPromise<DockerServerRuntimeInstance>()
+    val connection = ServerConnectionManager.getInstance().createTemporaryConnection(
+        RemoteServerImpl("DockerConnection", DockerCloudType.getInstance(), DockerCloudConfiguration.createDefault())
+    )
+    connection.connectIfNeeded(object : ServerConnector.ConnectionCallback<DockerDeploymentConfiguration> {
+        override fun errorOccurred(errorMessage: String) {
+            instancePromise.setError(errorMessage)
+        }
+
+        override fun connected(serverRuntimeInstance: ServerRuntimeInstance<DockerDeploymentConfiguration>) {
+            instancePromise.setResult(serverRuntimeInstance as DockerServerRuntimeInstance)
+        }
+    })
+
+    return instancePromise.await()
+}
+
+suspend fun getDockerApplicationRuntimeInstance(serverRuntime: DockerServerRuntimeInstance, imageId: String): DockerApplicationRuntime =
+    serverRuntime.findRuntimeLater(imageId, false).await()
+
+suspend fun pushImage(project: Project, dockerServerRuntime: DockerServerRuntimeInstance, ecrLogin: EcrLogin, pushRequest: EcrPushRequest) {
+    val (username, password) = ecrLogin
+    val model = DockerRepositoryModel().also {
+        val repoUri = pushRequest.remoteRepo.repositoryUri
+        it.registry = DockerRegistry().also { registry ->
+            registry.address = repoUri
+            registry.username = username
+            registry.password = password
+        }
+        it.repository = repoUri
+        it.tag = pushRequest.remoteTag
+    }
+
+    val dockerApplicationRuntime = getDockerApplicationRuntimeInstance(dockerServerRuntime, pushRequest.localImageId)
+    dockerApplicationRuntime.pushImage(project, model)
+}
+
+private const val NO_TAG_TAG = "<none>:<none>"
+internal fun Array<DockerAgentApplication>.toLocalImageList(): List<LocalImage> =
+    this.flatMap { image ->
+        image.imageRepoTags?.map { localTag ->
+            val tag = if (localTag == NO_TAG_TAG) null else localTag
```

Comment:
Up to you but maybe clearer?


`val tag = localTag.takeUnless { it == NO_TAG_TAG }`